### PR TITLE
WIP. PPLUS-480: Run composer update (for minor & patches)

### PR DESCRIPTION
**Pull request by VladislavStrelchenko**
Overview
Ticket: https://spryker.atlassian.net/browse/PPLUS-480

### CODE REVIEW

File: src/Upgrader/Communication/Command/UpgradeCommand.php [hunk 1]
```diff
@@ -58,6 +83,19 @@ public function execute(InputInterface $input, OutputInterface $output): int
...
+     *
+     * @return \Upgrader\Business\Upgrader\Request\UpgraderRequest
+     */
+    protected function createRequest(InputInterface $input): UpgraderRequest
+    {
+        $strategyOption = $input->getOption(static::OPTION_STRATEGY) ?? UpgradeStrategyEnum::COMPOSER_UPDATE;
+        $strategyEnum = new UpgradeStrategyEnum($strategyOption);
+
+        return new UpgraderRequest($strategyEnum);
+    }
```
Comment (kalaputsik) [inline src/Upgrader/Communication/Command/UpgradeCommand.php:97]: could we use "create*"" method only in the Factory?
we might remove createRequest method and create DTO in the "execute" method directly, without dependency on the UpgradeStrategyEnum

File: src/Upgrader/Business/UpgraderBusinessFactory.php [hunk 2]
```diff
@@ -55,9 +60,30 @@ class UpgraderBusinessFactory
     public function createUpgrader(): Upgrader
```
Comment (kalaputsik) [inline src/Upgrader/Business/UpgraderBusinessFactory.php:60]: we could set the Strategy inside UpgraderBusinessFactory

File: src/Upgrader/Business/Upgrader/Upgrader.php [hunk 3]
```diff
@@ -7,63 +7,75 @@
...
 use Upgrader\Business\VersionControlSystem\VcsInterface;
 
 class Upgrader implements UpgraderInterface
 {
     /**
-     * @var \Upgrader\Business\Upgrader\Bridge\ReleaseGroupTransferBridgeInterface
+     * @var \Upgrader\Business\Upgrader\Strategy\ComposerUpdateStrategy
      */
-    protected $releaseGroupManager;
+    protected $composerUpdateStrategy;
```
Comment (kalaputsik) [inline src/Upgrader/Business/Upgrader/Upgrader.php:21]: instead
```

/**
  * @var \Upgrader\Business\Upgrader\Strategy\ComposerUpdateStrategy
  */
 protected $composerUpdateStrategy;

 /**
  * @var \Upgrader\Business\Upgrader\Strategy\ReleaseGroupStrategy
  */
 protected $releaseGroupStrategy;

```
might be

```

/**
 * @var \Upgrader\Business\Upgrader\Strategy\UpgradeStrategyInterface
 */
protected $strategy;
```

File: src/Upgrader/Business/Upgrader/Upgrader.php [hunk 4]
```diff
@@ -7,63 +7,75 @@
...
-     * @param \Upgrader\Business\Upgrader\Bridge\ReleaseGroupTransferBridgeInterface $releaseGroupManager
-     * @param \Upgrader\Business\Upgrader\Bridge\PackageManagementSystemBridge $dataProviderManager
+     * @param \Upgrader\Business\Upgrader\Strategy\ComposerUpdateStrategy $composerUpdateStrategy
+     * @param \Upgrader\Business\Upgrader\Strategy\ReleaseGroupStrategy $releaseGroupStrategy
      * @param \Upgrader\Business\VersionControlSystem\VcsInterface $vcs
      */
     public function __construct(
-        ReleaseGroupTransferBridgeInterface $releaseGroupManager,
-        PackageManagementSystemBridge $dataProviderManager,
+        ComposerUpdateStrategy $composerUpdateStrategy,
```
Comment (kalaputsik) [inline src/Upgrader/Business/Upgrader/Upgrader.php:39]: our development should be based on the Interfaces, not on the realizations
in the *__construct*
we should use
\Upgrader\Business\Upgrader\Strategy\UpgradeStrategyInterface

File: src/Upgrader/Business/Upgrader/Upgrader.php [hunk 5]
```diff
@@ -7,63 +7,75 @@
...
             return $responses;
         }
-        $dataProviderResponse = $this->dataProviderManager->getNotInstalledReleaseGroupList();
-        $packageManagerResponses = $this->releaseGroupManager->requireCollection(
-            $dataProviderResponse->getReleaseGroupCollection()
-        );
-        $responses->addCollection($packageManagerResponses);
-        if (!$packageManagerResponses->hasSuccessfulResponse()) {
+
+        if ($request->getStrategyEnum()->isComposerUpdate()) {
```
Comment (kalaputsik) [inline src/Upgrader/Business/Upgrader/Upgrader.php:68]: might be
$upgradeResponses = $this->strategy->upgrade();

File: src/Upgrader/Business/PackageManager/PackageManager.php [hunk 6]
```diff
@@ -70,6 +81,32 @@ public function requireDev(PackageTransferCollection $packageCollection): Packag
         return $this->packageManagerClient->requireDev($packageCollection);
     }
 
+    /**
+     * @return \Upgrader\Business\PackageManager\Response\Collection\PackageManagerResponseCollection
+     */
+    public function update(): PackageManagerResponseCollection
+    {
+        $rawResponse = $this->packageManagerClient->update();
```
Comment (kalaputsik) [inline src/Upgrader/Business/PackageManager/PackageManager.php:89]: what about $packageManagerResponse ?

File: src/Upgrader/Business/PackageManager/PackageManager.php [hunk 7]
```diff
@@ -70,6 +81,32 @@ public function requireDev(PackageTransferCollection $packageCollection): Packag
...
+                    $rawResponse->getOutput(),
+                    $diffResponse->getPackageList(),
+                );
+                $updateResponseCollection->add($updateResponse);
+            }
+
+            return $updateResponseCollection;
+        }
+
+        return new PackageManagerResponseCollection([$rawResponse]);
```
Comment (kalaputsik) [inline src/Upgrader/Business/PackageManager/PackageManager.php:107]: why we are creating twice new PackageManagerResponseCollection(),
we could move it out of the "if" statement

File: src/Upgrader/Business/Upgrader/Strategy/ReleaseGroupStrategy.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    /**
+     * @var \Upgrader\Business\Upgrader\Bridge\PackageManagementSystemBridge
+     */
+    protected $dataProviderManager;
+
+    /**
+     * @param \Upgrader\Business\Upgrader\Bridge\ReleaseGroupTransferBridgeInterface $releaseGroupManager
+     * @param \Upgrader\Business\Upgrader\Bridge\PackageManagementSystemBridge $dataProviderManager
+     */
+    public function __construct(ReleaseGroupTransferBridgeInterface $releaseGroupManager, PackageManagementSystemBridge $dataProviderManager)
```
Comment (kalaputsik) [inline src/Upgrader/Business/Upgrader/Strategy/ReleaseGroupStrategy.php:30]: if we have an Interface, may we use it?

File: src/Upgrader/Business/Upgrader/Enum/UpgradeStrategyEnum.php [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+     */
+    public function isComposerUpdate(): bool
+    {
+        return $this->getValue() === static::COMPOSER_UPDATE;
+    }
+
+    /**
+     * @return bool
+     */
+    public function isReleaseGroup(): bool
```
Comment (kalaputsik) [inline src/Upgrader/Business/Upgrader/Enum/UpgradeStrategyEnum.php:35]: it's not used anywhere

File: src/Upgrader/Business/Upgrader/Enum/UpgradeStrategyEnum.php [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+    /**
+     * @var string
+     */
+    public const RELEASE_GROUP = 'release-group';
+
+    /**
+     * @return bool
+     */
+    public function isComposerUpdate(): bool
```
Comment (kalaputsik) [inline src/Upgrader/Business/Upgrader/Enum/UpgradeStrategyEnum.php:27]: will not be used anywhere

File: src/Upgrader/Communication/Command/UpgradeCommand.php (hunk 1 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Communication/Command/UpgradeCommand.php:97]: Deleted Enum

File: src/Upgrader/Business/UpgraderBusinessFactory.php (hunk 2 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/UpgraderBusinessFactory.php:60]: Set in UpdateStrategyGenerator

File: src/Upgrader/Business/Upgrader/Upgrader.php (hunk 3 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Upgrader/Upgrader.php:21]: Implemented by UpdateStrategyGenerator

File: src/Upgrader/Business/Upgrader/Upgrader.php (hunk 4 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Upgrader/Upgrader.php:39]: Thanks, replace by Interfaces

File: src/Upgrader/Business/Upgrader/Upgrader.php (hunk 5 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Upgrader/Upgrader.php:68]: ++

File: src/Upgrader/Business/Upgrader/Enum/UpgradeStrategyEnum.php (hunk 10 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Upgrader/Enum/UpgradeStrategyEnum.php:27]: deleted

File: src/Upgrader/Business/Upgrader/Enum/UpgradeStrategyEnum.php (hunk 9 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Upgrader/Enum/UpgradeStrategyEnum.php:35]: deleted

File: src/Upgrader/Business/Upgrader/Strategy/ReleaseGroupStrategy.php (hunk 8 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/Upgrader/Strategy/ReleaseGroupStrategy.php:30]: Fixed

File: src/Upgrader/Business/PackageManager/PackageManager.php (hunk 6 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/PackageManager/PackageManager.php:89]: renamed to $response

File: src/Upgrader/Business/PackageManager/PackageManager.php (hunk 7 above)
Comment (VladislavStrelchenko) [reply src/Upgrader/Business/PackageManager/PackageManager.php:107]: Fixed

**Pull request merged**